Assigning a write chan  to a read chan makes yaegi panic.

The following program `sample.go` triggers a panic:

```go
package main

func main() {
	var c chan<- struct{} = make(chan struct{})
	var d <-chan struct{} = c

	_ = d
}
```

Expected result:
```console
go run main.go 
# command-line-arguments
./main.go:5:6: cannot use c (type chan<- struct {}) as type <-chan struct {} in assignment
```

Got:
```console
yaegi main.go 
main.go:4:26: panic
reflect.Set: value of type chan<- struct {} is not assignable to type <-chan struct {}
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc000252600, 0x40)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0x9d
github.com/containous/yaegi/interp.(*Interpreter).Eval.func1(0xc000195de8)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/interp.go:333 +0xbc
panic(0xca8580, 0xc0000707b0)
	/usr/lib/go/src/runtime/panic.go:969 +0x166
github.com/containous/yaegi/interp.runCfg.func1(0xc0002940a0, 0xc000187100)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/run.go:115 +0x20c
panic(0xca8580, 0xc0000707b0)
	/usr/lib/go/src/runtime/panic.go:969 +0x166
reflect.Value.assignTo(0xca2900, 0xc0000101e8, 0x192, 0xddb5b8, 0xb, 0xca1ec0, 0x0, 0xc000189560, 0x12, 0xc000195be8)
	/usr/lib/go/src/reflect/value.go:2403 +0x426
reflect.Value.Set(0xca1ec0, 0xc000010200, 0x192, 0xca2900, 0xc0000101e8, 0x192)
	/usr/lib/go/src/reflect/value.go:1532 +0xbd
github.com/containous/yaegi/interp.assign.func4(0xc0002940a0, 0xc00000ee00)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/run.go:424 +0xb6
github.com/containous/yaegi/interp.runCfg(0xc000187100, 0xc0002940a0)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/run.go:121 +0x6d
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0001a61a0, 0xc000186400, 0xc000294000)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/run.go:99 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0001a61a0, 0xc000230b40, 0x85, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/interp.go:409 +0x52b
main.main()
	/home/jlevesy/code/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:149 +0x5d8
```

### Notes

This is probably a follow up to #702: we are able to detect during cfg a bunch of bugs related to channel assignments.

--- a/_test/assign15.go
+++ b/_test/assign15.go
@@ -1,0 +1,11 @@
+package main
+
+func main() {
+	var c chan<- struct{} = make(chan struct{})
+	var d <-chan struct{} = c
+
+	_ = d
+}
+
+// Error:
+// _test/assign15.go:5:26: cannot use type chan<- struct{} as type <-chan struct{} in assignment

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -518,6 +518,22 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 						err = src.cfgErrorf("invalid float truncate")
 						return
 					}
+
+					// TODO: Rudimentary type check at this point,
+					// improvements need to be made to make it better.
+					switch {
+					case dest.typ.untyped || src.typ.untyped:
+						// Both side of the assignment must be typed.
+					case isRecursiveField(dest):
+						// Recursive fields cannot be type checked.
+					case t0.Kind() == reflect.Interface || t0.Kind() == reflect.Func:
+						// We have no way of checking interfaces and functions.
+					case t1.AssignableTo(t0):
+						// All is well when they are assignable.
+					default:
+						err = src.cfgErrorf("cannot use type %s as type %s in assignment", src.typ.id(), dest.typ.id())
+						return
+					}
 				}
 				n.findex = dest.findex
 				n.level = dest.level
@@ -1157,6 +1173,7 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 				case n.ident == "iota":
 					n.rval = reflect.ValueOf(sc.iota)
 					n.kind = basicLit
+					n.typ.untyped = true
 				case n.ident == "nil":
 					n.kind = basicLit
 				case sym.kind == binSym:

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -524,8 +524,8 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 					switch {
 					case dest.typ.untyped || src.typ.untyped:
 						// Both side of the assignment must be typed.
-					case isRecursiveField(dest):
-						// Recursive fields cannot be type checked.
+					case isRecursiveType(dest.typ, dest.typ.rtype) || isRecursiveType(src.typ, src.typ.rtype):
+						// Recursive types cannot be type checked.
 					case t0.Kind() == reflect.Interface || t0.Kind() == reflect.Func:
 						// We have no way of checking interfaces and functions.
 					case t1.AssignableTo(t0):

--- a/interp/interp_consistent_test.go
+++ b/interp/interp_consistent_test.go
@@ -35,6 +35,7 @@ func TestInterpConsistencyBuild(t *testing.T) {
 		if filepath.Ext(file.Name()) != ".go" ||
 			file.Name() == "assign11.go" || // expect error
 			file.Name() == "assign12.go" || // expect error
+			file.Name() == "assign15.go" || // expect error
 			file.Name() == "bad0.go" || // expect error
 			file.Name() == "const9.go" || // expect error
 			file.Name() == "export1.go" || // non-main package

--- a/interp/type.go
+++ b/interp/type.go
@@ -1013,7 +1013,11 @@ func (t *itype) id() (res string) {
 	case arrayT:
 		res = "[" + strconv.Itoa(t.size) + "]" + t.val.id()
 	case chanT:
-		res = "<-" + t.val.id()
+		res = "chan " + t.val.id()
+	case chanSendT:
+		res = "chan<- " + t.val.id()
+	case chanRecvT:
+		res = "<-chan " + t.val.id()
 	case funcT:
 		res = "func("
 		for _, t := range t.arg {

